YKI(Frontend): OPHYKIKEH-150 Tarkistusarviointiin ilmoittautuminen, osa 2

pkoivisto opened the pull request:
## Yhteenveto

_Toteutuksen yhteenveto..._

## Lisätiedot

_Tehdyt muutokset..._

## Huomautukset

_Muut mahdolliset huomautukset..._

## Check-lista

- [ ] yarn.lock päivitetty
- [ ] Käännösten Excel-tiedosto päivitetty
- [ ] UI-muutoksia testattu eri selaimilla ja mobiilinäkymässä

pkoivisto committed:
SHARED(Frontend): Support arbitrary React node as DialogContent

pkoivisto committed:
YKI(Frontend): Redirect to payment or display error dialog when evaluation order is submitted

pkoivisto committed:
YKI(Frontend): Implement redux, saga, serialization etc. logic for submitting evaluation order

pkoivisto committed:
YKI(Frontend): Improve error indicators, disable navigation protection when redirecting to Paytrail

pkoivisto committed:
YKI(Frontend): Simplify redirect flow, add evaluation order status page to redirect to from Paytrail [deploy]

jrkkp commented on frontend/packages/yki/src/components/reassessment/evaluationOrder/PublicEvaluationOrderForm.tsx:
```diff
@@ -241,19 +259,107 @@ const AcceptConditions = () => {
+        <div>
+          <Text>{t('fixErrors')}</Text>
+          <ul>
+            {errors.map((error, i) => (
+              <li key={i}>
+                <Text>{t(`errors.${error}`)}</Text>
+              </li>
```
Tässä vois käyttää keynä suoraan tuota erroria. indexi kai ongelmallinen kun järjestys voi muuttua ja sillon sama key viittais eri sisältöön.

pkoivisto replied:
Joo, voisin muuttaa.

pkoivisto committed:
YKI(Frontend): Avoid using array index as key

mikhuttu commented on frontend/packages/yki/src/enums/api.ts:
```diff
@@ -3,4 +3,6 @@ export enum APIEndpoints {
   Evaluations = '/yki/api/evaluation',
   Evaluation = '/yki/api/evaluation/:evaluationId',
+  EvaluationOrder = '/yki/api/evaluation/:evaluationId/order',
+  EvaluationPaymentRedirect = '/yki/api/evaluation-payment/v2/:evaluationOrderId/redirect',
```
Jäi käyttämättömäksi

pkoivisto replied:
Totta, poistan.

mikhuttu commented on frontend/packages/yki/src/pages/EvaluationOrderStatusPage.tsx:
```diff
@@ -0,0 +1,69 @@
+import { Box } from '@mui/system';
```
Taitaa olla sama komponentti mutta muualla importattu kai `@mui/material` alta

pkoivisto replied:
On noita näköjään parissa muussakin paikassa haettu `@mui/system` alta, mutta tosiaan huomattavan paljon enemmän `@mui/material` alta. Vaihdetaan tuohon toiseen.

mikhuttu commented on frontend/packages/yki/src/pages/EvaluationOrderStatusPage.tsx:
```diff
@@ -0,0 +1,69 @@
+};
+
+const useOrderStatusComponent = () => {
+  const [params] = useSearchParams();
+  const status = params.get('status');
```
Nää olis ehkä loogisempia tuolla ylemmällä tasolla, jos tää metodi olisi tyyppiä `getContent = (status: OrderStatus)` tms.

mikhuttu commented on frontend/packages/yki/src/utils/serialization.ts:
```diff
@@ -69,4 +76,49 @@ export class SerializationUtils {
+    return {
+      first_names: payerDetails.firstNames as string,
+      last_name: payerDetails.lastName as string,
+      birthdate: DateUtils.serializeDate(payerDetails.birthdate) as string,
+      email: payerDetails.email as string,
```
Nämä vois vaihtaa päittäin, kun ovat toisin päin EvaluationOrderRequest:n alla

Huomasin muuten testailemalla myös että formin voi yrittää lähettää ei-validilla syntymäajalla jolloin käyttäjälle näytetään geneerinen virheilmoitus. Toisena huomiona että email regex on backendillä eri kuin frontendissä ja on mahdollista postata formi jossa backendin mielestä ei-validi email.

pkoivisto replied:
Tän osalta veikkaan että voisi muokata tuota bäkkärin regexiä vastaamaan tätä sharedin toteutusta ja tehdä vastaavat muutokset sitten myös nykyisen frontendin osalta. Tarttee ehkä perehtyä vielä uudempaan kertaan siihen bäkkärin regexiin varmuuden vuoksi, mutta kyllähän nää pitäisi jotenkin synkata keskenään.

mikhuttu commented on frontend/packages/yki/src/redux/reducers/evaluationOrder.ts:
```diff
@@ -48,8 +53,12 @@ const evaluationOrderSlice = createSlice({
+      action: PayloadAction<EvaluationOrderResponse>
+    ) {
       state.submitOrderState = APIResponseStatus.Success;
+      state.evaluationPaymentRedirectResponse = action.payload;
```
Mitä jos tallettaisi vain redirectin esim. muuttujaan `evaluationPaymentRedirect?: string` ?

pkoivisto replied:
Käy tämäkin. Muutettu.

mikhuttu commented on frontend/packages/yki/src/pages/EvaluationOrderPage.tsx:
```diff
@@ -75,6 +80,11 @@ export const EvaluationOrderPage = () => {
       });
```
Tää error-keissin handlaus ei oo toteutettu kovin hyvin, tai ainakaan virheilmoitusta ei näytetä kun `toasts.notFound` puutuu sivun lokalisaatioista. Meiltä näkyy puuttuvan myös handlaus tilanteeseen, jossa tarkistusarviointia ei löydy urlissa olevalla id:llä.

Vaikuttais siltä että ehkä tuo riviltä 58 alkava `resetStateOnUnmount` on tässä ongelmana, sillä virhetilanteen handlaamisen sijasta (loadPeriodState tilassa Error) tuo päätyy tilaan NotStarted? Korjaisko tässä PR:ssä tämän jo?

pkoivisto replied:
Parantelin tätä virhetilanteiden käsittelyä, mielestäni tää nyt toimii kohtuu järkevästi. Poistin myös uudelleenohjauksen listaussivulle, mielestäni selkeämpää kuitenkin että näytetään sen sijaan virheviesti.

mikhuttu reviewed (commented):
Vaikuttaa varsin toimivalta jo. Virhetilanteisiin liittyen tuli joitakin huomioita tehtyä. Korjaako näitä sitten tän PR:n yhteydessä vai omana pull requestinaan niin toinen juttu.

pkoivisto committed:
YKI(Frontend): Review fixes

mikhuttu reviewed (approved):
Tää ok jos virhetilanteiden käsittelyt haluaa tehdä omana tikettinään siirtäen ne sinne

pkoivisto committed:
YKI(Frontend): Add missing header translation

pkoivisto committed:
YKI(Frontend): Improve error case handlin [deploy]

Don't attempt to navigate away from evaluation order page in case an API error occurs.
Instead show a simple error message.
Eschew raising a generic error from the corresponding saga, instead raise an error toast from within the page itself.

mikhuttu commented on frontend/packages/yki/src/pages/EvaluationOrderPage.tsx:
```diff
@@ -98,27 +84,54 @@ export const EvaluationOrderPage = () => {
+                {t('title')}
+              </H1>
+              <HeaderSeparator />
+              <H2>Tarkistusarvionnin tietoja ei löytynyt!</H2>
```
Laittaisko tän tulemaan vaikka lokalisoinneista? Tai korjataanko ylipäänsä tätä toteutusta kun Emmalta saa kommentteja tähän?

pkoivisto replied:
Oho, joo. Tää oli näköjään vielä jäänyt inlineksi, fiksaan. Varmasti kannattaa vielä Emman kautta pyöräyttää sitten jossain vaiheessa.

mikhuttu reviewed (commented):
Olitko muuten deployaamassa tätä vielä untuvalle? Jos testais siellä tätä virhekäsittelyä vielä

pkoivisto commented:
> Olitko muuten deployaamassa tätä vielä untuvalle? Jos testais siellä tätä virhekäsittelyä vielä

Juu, vien juuri untuvalle.

pkoivisto committed:
YKI(Frontend): Convert inline text to translation

pkoivisto committed:
YKI(Frontend): Improve birthdate validation
